feat(extensions): Add extension permissions

debsmita1 opened the pull request:
## Hey, I just made a Pull Request!

**Resolves:**

https://issues.redhat.com/browse/RHIDP-6766
     
**How to test:**

Install rbac frontend and backend plugins
Add the Github signin page
Add the following in you rbac-policy.csv file
```
p, role:default/admins, extension-plugin, read, allow
p, role:default/admins, extension-plugin, create, allow
```

Add the following to your `app-config.yaml`
```
permission:
  enabled: true
  rbac:
    pluginsWithPermission:
      - marketplace
      - catalog
      - permission
    policies-csv-file: <path-to-your-csv-file>/rbac-policy.csv
    policyFileReload: true
    admin:
      users:
        - name: user:default/<github-id>
        - name: group:default/admins
  ```
  
  
  **Screenshots/GIF:**
  
User with no config read permission
  
![Screenshot 2025-04-29 at 7 10 52 PM](https://github.com/user-attachments/assets/6394e849-84f7-4ec4-af42-e37c08871d72)

User with only config read permission

https://github.com/user-attachments/assets/f758f8cb-910c-49af-9f4c-249157ae0a59


User with both read and create permissions

https://github.com/user-attachments/assets/af8bd9be-b046-4300-b92f-1e8647c53ec2


User with conditional permissions

https://github.com/user-attachments/assets/77c371fc-9425-4227-a6aa-4638dfe66d24


TODO (to be addressed in separate PRs): 
- Change "Extension" to "Extensions"
- Update RBAC documentation to include the newly added Extensions permissions
- Update Extensions frontend plugin readme to include the policies to be added when the  Permission framework is enabled

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/redhat-developer/rhdh-plugins/blob/main/CONTRIBUTING.md#creating-changesets))
- [ ] Added or Updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

ShiranHi commented:
Thank you @debsmita1, it looks great!

I have two requests: 
1. **For user with no config read permission:** let's change the tooltip to be "You don't have permission to install plugins or view their configurations. Contact your administrator to request access or assistance."
2. **For user with only config read permission:** let's keep the Install button there, but disabled, and add a tooltip on hover "You don't have permission to install plugins or edit their configurations. Contact your administrator to request access or assistance."

debsmita1 commented:
@ShiranHi 
> 2. **For user with only config read permission:** let's keep the Install button there, but disabled, and add a tooltip on hover "You don't have permission to install plugins or edit their configurations. Contact your administrator to request access or assistance."

Are you suggesting that we remove the 'View' button and instead display a disabled 'Install' button? The idea is to allow users with read-only permissions to access the plugin configuration.

logonoff commented on workspaces/marketplace/plugins/marketplace/src/components/MarketplacePluginContent.tsx:
```diff
@@ -188,6 +195,42 @@ export const MarketplacePluginContent = ({
+      <Tooltip
+        title={
+          canInstallPlugin.data?.status === 'Unauthorized'
+            ? 'You donot have the necessary permissions to read the plugin configuration. Please contact your administrator for access rights.'
```
```suggestion
            ? 'You don't have the necessary permissions to read the plugin configuration. Please contact your administrator for access rights.'
```

PatAKnight commented on workspaces/marketplace/plugins/marketplace/src/hooks/useExtensionReadConfigPermission.ts:
```diff
@@ -0,0 +1,40 @@
+  });
+};
+
+export const useExtensionInstallPluginPermission = () => {
+  const canInstallPlugin = usePermission({
+    permission: extensionPluginCreatePermission,
+    resourceRef: extensionPluginCreatePermission.resourceType,
+  });
```
We recently found out that we have been using `usePermission` incorrectly. The `resourceRef` is suppose to point to the resource in question that you are doing the permission check against. See what I had to do for the [RBAC frontend plugin](https://github.com/backstage/community-plugins/blob/main/workspaces/rbac/plugins/rbac/src/components/RolesList/DeleteRole.tsx#L46) recently.

PatAKnight commented on workspaces/marketplace/plugins/marketplace-backend/src/router.ts:
```diff
@@ -108,6 +159,57 @@ export async function createRouter({
+  router.get('/plugin/:namespace/:name/configuration', async (req, res) => {
+    const [readDecision, installDecision] = await Promise.all([
+      authorizeConditional(req, extensionPluginReadPermission),
+      authorizeConditional(req, extensionPluginCreatePermission),
```
I wonder if this should just be the read permission only and then we can use the create permission for `post`.

PatAKnight reviewed (commented):
Left a couple of initial comments but will need to do a deeper dive into the plugin itself to try to understand it a bit better. If you want, we can sync up on this as I recently had to do this for the [RBAC plugin](https://github.com/backstage/community-plugins/pull/3265) and it is still fairly top of mind.

debsmita1 commented:
> > Are you suggesting that we remove the 'View' button and instead display a disabled 'Install' button? The idea is to allow users with read-only permissions to access the plugin configuration.
> 
> Sorry for the confusion! We have two "Install" buttons. Let me clarify it:
> 
> 1. **For user with no config read permission:** let's change the tooltip text that appears **on the drawer** when hovering over the Install button to: "You don't have permission to install plugins or view their configurations. Contact your administrator to request access or assistance."
> 2. **For user with only config read permission:** let's keep the Install button **on the configuration page**, but disabled, and add a tooltip on hover "You don't have permission to install plugins or edit their configurations. Contact your administrator to request access or assistance."

@ShiranHi  Addressed your comments and updated the GIF

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace-backend/src/router.ts:
```diff
@@ -108,6 +159,112 @@ export async function createRouter({
+      extensionPluginReadPermission,
+    );
+    if (readDecision.result === AuthorizeResult.DENY) {
+      res.status(403).json({ error: 'Forbidden' });
+      return;
```
Can we throw an error instead? For all 403 cases.

```tsx
import {
  NotAllowedError,
} from '@backstage/errors';
```

```suggestion
        throw new NotAllowedError(`Not allowed to configure ${req.params.namespace}:${req.params.name}`);
```

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace-common/src/permissions.ts:
```diff
@@ -0,0 +1,148 @@
+/** This permission grants access to the endpoint that reads the configuration of the extension plugin
+ * @public
+ */
+export const extensionPluginReadPermission = createPermission({
```
The variable name should include Configuration as well, to avoid confusing that this is a permission to read the plugin metadata.

IMHO, as long as we don't differentiate between Create update and Delete we should have two permissions:

* read
* write or update

If you want have separate for enableing and disabling the plugin I personally would use

* enable
* disable

Actually, that's then maybe not a configuration enable/disable, its then maybe `'extensions.plugin.enable`.

But I think we could start here also with `extensions.plugin.configuration.write`, because it just does this at the moment.

christoph-jerolimov commented on workspaces/marketplace/plugins/marketplace/src/hooks/useExtensionConfigPermission.ts:
```diff
@@ -0,0 +1,36 @@
+
+import { useMarketplaceApi } from './useMarketplaceApi';
+
+export const useExtensionConfigPermission = (
```
other hook names are aligned with the MarketplaceApi function they call and just replace get with use

christoph-jerolimov reviewed (commented):
Some thoughts, please take a look. I'm fine if we want align some of these also later:

christoph-jerolimov reviewed (approved):
As discussed, approving this spike but it would be nice if we can address some reviews soon. Esp. renaming all permission names to `extensions.*` before we release the plugin.

I will set the release on hold until this changes was done!

debsmita1 commented:
> As discussed, approving this spike but it would be nice if we can address some reviews soon. Esp. renaming all permission names to `extensions.*` before we release the plugin.
> 
> I will set the release on hold until this changes was done!

Addressed the review comments here https://github.com/redhat-developer/rhdh-plugins/pull/762
